servers/src/main/java/tachyon/master/next/IndexSet.java

```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  /**
+   * Add an object to the set.
+   *
+   * @param object the object to add
+   */
+  public void add(T object) {
+    mSet.add(object);
+
+    for (String field : mFields.keySet()) {
+      Map<Object, Set<T>> fieldValueToSet = mSetIndexedByFieldValue.get(field);
+      Object value = getField(object, field);
+      if (fieldValueToSet.containsKey(value)) {
+        fieldValueToSet.get(value).add(object);
+      } else {
+        Set<T> set = new HashSet<T>();
+        set.add(object);
+        fieldValueToSet.put(value, set);
+      }
+    }
+  }
+
+  /**
+   * @return a set of all objects
+   */
+  public Set<T> all() {
+    return mSet;
+  }
+
+  /**
+   * Whether there is an object with the specified field value in the set.
+   *
+   * @param fieldName the field name
+   * @param value the field value
+   * @return true if there is one such object, otherwise false
+   */
+  public boolean contains(String fieldName, Object value) {
+    return !get(fieldName, value).isEmpty();
+  }
+
+  /**
+   * Get a subset of objects with the specified field value.
+   *
+   * @param fieldName the field name
+   * @param value the field value to be satisfied
+   * @return the set of objects or an empty set if no such object exists
+   */
+  public Set<T> get(String fieldName, Object value) {
+    Set<T> set = mSetIndexedByFieldValue.get(fieldName).get(value);
+    return set == null ? new HashSet<T>() : set;
+  }
+
+  /**
+   * Get an object with the specified field value from the set. Assumption is, in the set of
+   * objects, only one object has the field value.
+   *
+   * @param fieldName the field name
+   * @param value the field value
+   * @return the object or null if there is no such object
+   */
+  public T getSingle(String fieldName, Object value) {
+    Set<T> all = get(fieldName, value);
+    return all.isEmpty() ? null : all.iterator().next();
+  }
+
+  /**
+   * Remove an object from the set.
+   *
+   * @param object the object to remove
+   * @return true if removed successfully, otherwise false
+   */
+  public boolean remove(T object) {
+    boolean success = mSet.remove(object);
+
+    for (String field : mFields.keySet()) {
+      Object fieldValue = getField(object, field);
+      Set<T> set = mSetIndexedByFieldValue.get(field).remove(fieldValue);
+      if (set != null) {
+        success = success && set.remove(object);
+        if (!set.isEmpty()) {
+          mSetIndexedByFieldValue.get(field).put(fieldValue, set);
+        }
+      }
+    }
+
+    return success;
+  }
+
+  /**
+   * Remove the subset of objects with the specified field value.
+   *
+   * @param fieldName the field name
+   * @param value the field value
+   * @return the set of objects that are removed, if no object is removed, the returned set is empty
+   * @return true if the objects are removed, otherwise false
+   */
+  public boolean remove(String fieldName, Object value) {
```

Comment:
In general, we overload two functions to the same name **ONLY IF** they provide exactly the same functions except taking parameters in different types. Here is not the case for `remove`. How about `removeByField`